src/traceme.c

```diff
@@ -0,0 +1,431 @@
+ */
+static void
+poll_loop(struct tracer_ctx *tr_ctx)
+{
+    __u64                       head;
+    struct perf_event_mmap_page *mmap_header = tr_ctx->mmap_header;
+    int                         n_events = 0;
+    size_t                      num_wakes = 0;
+    struct pollfd               pfds[2] = {
+        {tr_ctx->perf_fd,           POLLIN | POLLHUP,   0},
+        {tr_ctx->loop_done_fds[0],  POLLHUP,            0}
+    };
+
+    while (1) {
+        n_events = poll(pfds, 2, INFTIM);
+        if (n_events == -1) {
+            err(EXIT_FAILURE, "poll");
+        }
+
+        if ((pfds[0].revents & POLLIN) || (pfds[1].revents & POLLHUP)) {
+            /* See <linux/perf_event.h> for why we need the asm block */
+            head = mmap_header->aux_head;
+            asm volatile ("" : : : "memory");
+
+            /* We were awoken to read out trace info, or we tracing stopped */
+            num_wakes++;
+            DEBUG("wake");
+            DEBUG("aux_head=  0x%010llu", head);
+            DEBUG("aux_tail=  0x%010llu", mmap_header->aux_tail);
+            DEBUG("aux_offset=0x%010llu", mmap_header->aux_offset);
+            DEBUG("aux_size=  0x%010llu", mmap_header->aux_size);
+
+            read_circular_buf(tr_ctx->aux, mmap_header->aux_size,
+                head, &mmap_header->aux_tail, tr_ctx->out_fd);
+
+            if (pfds[1].revents & POLLHUP) {
+                break;
+            }
+        }
+
+        if (pfds[0].revents & POLLHUP) {
+            break;
+        }
+    }
+
+    DEBUG("poll loop exit: awoke %zu times", num_wakes);
+}
+
+/*
+ * Set up Intel PT buffers and start a poll() loop for reading out the trace.
+ */
+static void *
+tracer_thread(void *arg)
+{
+    struct perf_event_attr      attr;
+    struct perf_event_mmap_page *header;
+    void                        *base, *data, *aux;
+    int                         page_size;
+    struct tracer_ctx           *tr_ctx = (struct tracer_ctx *) arg;
+    char                        pt_type_str[MAX_PT_TYPE_STR];
+    FILE                        *pt_type_file;
+
+    DEBUG("tracer init");
+
+    /* Basic configuration */
+    memset(&attr, 0, sizeof(attr));
+    attr.size = sizeof(attr);
+    attr.size = sizeof(struct perf_event_attr);
+
+    /* Get the perf "type" for Intel PT */
+    pt_type_file = fopen(SYSFS_PT_TYPE, "r");
+    if (pt_type_file == NULL) {
+        err(EXIT_FAILURE, "fopen");
+    }
+    if (fgets(pt_type_str, sizeof(pt_type_str), pt_type_file) == NULL) {
+        err(EXIT_FAILURE, "fgets");
+    }
+    attr.type = atoi(pt_type_str);
+
+    /* Exclude the kernel */
+    attr.exclude_kernel = 1;
+
+    /* Exclude the hyper-visor */
+    attr.exclude_hv = 1;
+
+    /* Start disabled */
+    attr.disabled = 1;
+
+    /* No skid */
+    attr.precise_ip = 3;
+
+    /* Acquire file descriptor through which to talk to Intel PT */
+    tr_ctx->perf_fd = syscall(SYS_perf_event_open, &attr,
+        tr_ctx->target_pid, -1, -1, 0);
```

Comment:
Shouldn't this be indented to just after the open bracket?

Reply:
Nah, that's PEP8-style.